### PR TITLE
AP_Scripting: path generation, closed loop trajectory tracking implem…

### PR DESCRIPTION
**Pull request by MatthewHampsey**
…ented

Pair programmed with @tridge and Andy Palmer

### CODE REVIEW

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 1]
```diff
@@ -141,6 +142,19 @@ userdata Vector2f operator +
 userdata Vector2f operator -
 userdata Vector2f method copy Vector2f
 
+userdata Quaternion field q1 float'skip_check read write
```
Comment (IamPete1) [inline libraries/AP_Scripting/generator/description/bindings.desc:145]: needs docs, generate with `./waf copter --scripting-docs` Then copy out the new ones and put them into the old docs. Bonus points for also adding useful param names and descriptions.

Comment (MatthewHampsey) [reply libraries/AP_Scripting/generator/description/bindings.desc:145]: Will do, cheers.

Comment (timtuxworth) [reply libraries/AP_Scripting/generator/description/bindings.desc:146]: This seems interesting. What does it do exactly?

**Comment by tridge** [comment]
a few things needed:
 - a README.md in the Trajectory/ directory
 - some sample missions (for CMAC)
 - an autotest that tests one of the missions

File: libraries/AP_Math/quaternion.cpp [hunk 2]
```diff
@@ -534,6 +534,19 @@ void QuaternionT<T>::from_axis_angle_fast(const Vector3<T> &axis, T theta)
     q4 = axis.z * st2;
 }
 
+template <typename T>
```
Comment (tridge) [inline libraries/AP_Math/quaternion.cpp:537]: comment that explains what this does

File: libraries/AP_Math/quaternion.h [hunk 3]
```diff
@@ -112,6 +112,8 @@ class QuaternionT {
     // the axis vector must be length 1, theta should less than 0.17 radians (i.e. 10 degrees)
     void        from_axis_angle_fast(const Vector3<T> &axis, T theta);
 
+    void        from_angular_velocity(const Vector3<T>& angular_velocity, float time_delta);
```
Comment (tridge) [inline libraries/AP_Math/quaternion.h:115]: comment here too

File: libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua [hunk 4]
```diff
@@ -0,0 +1,1102 @@
+--[[ perform simple aerobatic manoeuvres in AUTO mode
+cmd = 1: axial rolls, arg1 = roll rate dps, arg2 = number of rolls
```
Comment (tridge) [inline libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua:2]: comments need updating on what fns are available

File: libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua [hunk 5]
```diff
@@ -0,0 +1,1102 @@
...
+
+--Just used this to test the above function, can probably delete now.
+function test_angular_rate()
+   local vector1 = makeVector3f(1.0, 0.0, 0.0)
+   local vector2 = makeVector3f(1.0, 1.0, 0.0)
+   local angular_rate = vectors_to_angular_rate(vector1, vector2, 1.0)
+   gcs:send_text(0, string.format("angular rate: %.1f %.1f %.1f", math.deg(angular_rate:x()), math.deg(angular_rate:y()), math.deg(angular_rate:z())))
+end
+
+--test_angular_rate()
```
Comment (tridge) [inline libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua:792]: remove commented out code

File: libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua [hunk 6]
```diff
@@ -0,0 +1,1102 @@
...
+   path_var.positions_ef[1] = next_target_pos_ef:copy()
+
+   --roll angle relative to maneuver position without rolling
+   path_var.roll_angles_bf[-1] = path_var.roll_angles_bf[0]
+   path_var.roll_angles_bf[0] = path_var.roll_angles_bf[1]
+   path_var.roll_angles_bf[1] = next_target_angle
+
+   local current_measured_pos_ef = ahrs:get_relative_position_NED_origin()
+
+   -- local path_error = {}
```
Comment (tridge) [inline libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua:939]: remove commented out code

File: libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua [hunk 7]
```diff
@@ -0,0 +1,1102 @@
...
+   local total_ang_vel_bf = ahrs:earth_to_body(total_ang_vel_ef)
+
+   local mf_axis = makeVector3f(1, 0, 0)
+   path_var.accumulated_orientation_rel_ef:earth_to_body(mf_axis)
+
+   local orientation_rel_mf_with_roll_angle = Quaternion()
+   orientation_rel_mf_with_roll_angle:from_axis_angle(mf_axis, path_var.roll_angles_bf[0])
+   orientation_rel_ef_with_roll_angle = orientation_rel_mf_with_roll_angle*path_var.accumulated_orientation_rel_ef
+
+   --logger.write("ACCO",'r,p,y', 'fff', orientation_rel_ef_with_roll_angle:get_euler_roll(), orientation_rel_ef_with_roll_angle:get_euler_pitch(), orientation_rel_ef_with_roll_angle:get_euler_yaw())
```
Comment (tridge) [inline libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua:1006]: remove commented out code

**Review by tridge — CHANGES_REQUESTED** [review]

**Pull request merged**